src/tests/disjoint_set_property_test.cpp

```diff
@@ -0,0 +1,301 @@
+static void check_max_rank(souffle::DisjointSet& ds, souffle::PiggyList<block_t>& blocks) {
+    testutil::ignore(ds);
+#ifdef _OPENMP
+#pragma omp parallel for
+#endif
+    for (std::size_t i = 0; i < blocks.size(); ++i) {
+        auto b = blocks.get(i);
+        LOCAL_ASSERT_LT(DisjointSet::b2r(b), blocks.size());
+    }
+}
+
+// `pr2b` undoes `b2r` and `b2p`.
+static void check_pr2b_undoes_b2r_b2p(souffle::DisjointSet& ds, souffle::PiggyList<block_t>& blocks) {
+    testutil::ignore(ds);
+#ifdef _OPENMP
+#pragma omp parallel for
+#endif
+    for (std::size_t i = 0; i < blocks.size(); ++i) {
+        auto b = blocks.get(i);
+        LOCAL_ASSERT_EQ(DisjointSet::pr2b(DisjointSet::b2p(b), DisjointSet::b2r(b)), b);
+    }
+}
+
+// Check invariants
+static void check(souffle::DisjointSet& ds, souffle::PiggyList<block_t>& blocks) {
+    LOCAL_ASSERT_EQ(ds.size(), blocks.size());
+    check_parent_rank(ds, blocks);
+    check_max_rank(ds, blocks);
+    check_pr2b_undoes_b2r_b2p(ds, blocks);
+}
+
+// Perform some number of randomized API calls to produce a random disjoint set.
+static souffle::PiggyList<block_t> randomize_disjoint_set(
+        souffle::DisjointSet& ds, std::size_t min_size, std::size_t max_size) {
+    assert(max_size > 0);
+    souffle::PiggyList<block_t> blocks;
+    auto random_node = [&blocks] { return DisjointSet::b2p(blocks.get(random() % blocks.size())); };
+    // Must be non-empty
+    blocks.append(ds.makeNode());
+#ifdef _OPENMP
+#pragma omp parallel for
+#endif
+    for (std::size_t i = 0; i < (random() % max_size) + min_size; ++i) {
+        // Insertions should be 2x as likely as other operations so we get
+        // reasonably sized sets with more than 1 equivalence class.
+        if (random() % 2) {
+            blocks.append(ds.makeNode());
+        } else {
+            if (random() % 2) {
+                ds.unionNodes(random_node(), random_node());
+            } else {
+                // Other mutating methods
+                switch (random() % 3) {
+                    case 0: {
+                        ds.findNode(random_node());
+                        break;
+                    }
+                    case 1: {
+                        ds.sameSet(random_node(), random_node());
+                        break;
+                    }
+                    default: {
+                        // Non-mutating methods (queries) - no need to perform these
+                        // often, they can all be about as likely as one another.
+                        if (random() % 2) {
+                            ds.size();
+                        } else {
+                            ds.get(random_node());
+                        }
+                        break;
+                    }
+                }
+            }
+        }
+    }
+    while (blocks.size() < min_size) {
+        blocks.append(ds.makeNode());
+    }
+    check(ds, blocks);
```

Comment:
Oh, I think this line doesn't do what I hoped it would... Since I'm inserting each node into `blocks` just after it's been created, they won't be up-to-date with `ds.a_blocks`... Would it be OK to expose a read-only iterator over `a_blocks`? Or some other kind of operation that would expose the collection of blocks in the `DisjointSet`?

If that's not permissible, I can just leave out `check` and friends.